[CI] don't run jobs on draft PRs

ZuleykhaPavlichenkova-TomTom opened the pull request:
We'd like to skip running CI on draft PRs and stop running CI on the PRs that got drafted.

Expected behaviour:

| Event | Current behaviour | Intended behaviour of CI after PR merged |
|---|---|---|
| Open PR | CI starts | CI starts |
| Open **draft** PR | CI starts | CI starts, runs the `check-draft` job and skips all other jobs |
| Convert to draft | - | CI starts, runs the `check-draft` job and skips all other jobs; previously triggered CI's got cancelled because of the same concurrency group |
| Ready for review | CI starts | CI starts; previously triggered CI's got cancelled because of the same concurrency group |
| Push to open PR | `DraftPR.yml` turns it into a draft; previous CI keeps running | CI starts; `DraftPR.yml` turns it into a draft; then it becomes the same as `Convert to draft` |
| Push to **draft** PR | - | CI starts, runs the `check-draft` job and skips all other jobs |

Expected effect of the PR is stop running CI on branches that are outdated or not ready to be evaluated yet.

ZuleykhaPavlichenkova-TomTom committed:
don't run jobs on PR drafts

ZuleykhaPavlichenkova-TomTom committed:
skip all jobs in Main.yml of PR draft

ZuleykhaPavlichenkova-TomTom committed:
we'd like to skip running CI on draft PRs that's why if: github.event.pull_request.draft == false condition was added to each job in workflows triggered on PR

ZuleykhaPavlichenkova-TomTom committed:
clean up if conditions and added condition to skip running Regression.yml on PR drafts

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into rearrange-ci

carlopi reviewed (changes requested):
I think centralizing the checks would make more sense, and while discussing with @hmeriann we found a solution also to the wider problem of cancelling now drafted PRs.

ZuleykhaPavlichenkova-TomTom committed:
add job check-draft to create an output based on value of which workflow will decide run or skip all other jobs.
add a PR case converted_to_draft, because we want be able to cancel CI runs on drafted PR immediately

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'rearrange-ci' of https://github.com/hmeriann/duckdb into rearrange-ci

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into rearrange-ci

ZuleykhaPavlichenkova-TomTom committed:
add dependency on check-draft job to the linux-config job

ZuleykhaPavlichenkova-TomTom committed:
add accidentially removed sondition to the Valgrind job

carlopi commented on .github/workflows/CodeQuality.yml:
```diff
@@ -39,10 +39,21 @@ env:
   GH_TOKEN: ${{ secrets.GH_TOKEN }}
 
 jobs:
+  check-draft:
+    name: Don't trigger further jobs to run on draft PR
+    runs-on: ubuntu-24.04
+    outputs:
+      is_draft: ${{ steps.check-draft.outputs.is_draft }}
+    steps:
+    - id: check-draft
+      run: |
+        echo "is_draft=${{ github.event.pull_request.draft }}" >> $GITHUB_OUTPUT
+
```
Question: could this be somehow made so that the other steps only have a `check-draft` ?

I think it should be possible (but would need to be tested), using your previous idea, something like:
```
  check-draft:
     name: ....
     if: ${{ github.event.pull_request.draft != true }}
     runs-on: ubuntu-24.04
     steps:    ## are those even needed
     - id: some_id
        run: echo "This is not a draft"
```

hmeriann replied:
Actually yes, thank you for the suggestion!

ZuleykhaPavlichenkova-TomTom committed:
remove dependency from chained jobs

ZuleykhaPavlichenkova-TomTom committed:
simplify chck-draft job

carlopi reviewed (commented):
I think this should work, and make stuff way nicer, thanks a lot!

I have NOT personally tried with this on a fork, and I am unsure of what happens in some corner cases, so I would probably recommend against merging until say Monday (by then there will be some more testing in, AND bandwidth to actual fix issues we hit on the go). In particular one question is: what happens on non pull_requests? Say one has a branch in its own fork, would CI run or not?

---

One comment independent from this PR, especially on workflows, changing existing code just to format it increases the diff AND consequently the load on reviewers. I get once stuff is formatted, it will stay so, so it has even a positive impact, but it's a bit confusing on doing a review.

ZuleykhaPavlichenkova-TomTom committed:
it's not valid without steps

hmeriann commented:
> it increases the diff

oh! didn't think about that this time. I can revert and recommit

ZuleykhaPavlichenkova-TomTom committed:
don't skip running on all other event names

carlopi committed:
Fixup comparsion

ZuleykhaPavlichenkova-TomTom committed:
cleaning up, add a comment

carlopi reviewed (approved):
Thanks, I think this makes a lot of sense, and should remove quite some unnecessary work that is currently run up to conclusion when already a draft or outdated state.

hmeriann commented:
There is a minimal example on which I've tested the condition not in the duckdb/duckdb fork https://github.com/zuleykhapa/nbc/pull/10/

Mytherin commented:
I wonder if it makes sense to have an "orchestrator" task that launches subsequent actions? That seems like a more clean approach versus the current approach of having to remember to add `-needs: ...` to every action.

Mytherin commented:
Thanks!
